WT-10353 Implement deleteObject() in connection class for GCP

sonar3ddy opened the pull request:
Implemented delete_object() in gcp connection class which returns 0 on success and -1 if there is an error. The status code is printed if an error occurs.

The delete_object function has been changed to be not const as the gcp client method DeleteObject that is called within the function is not const.
object_key has been renamed to be object_name to match gcp method naming.

The delete_object function has been updated to include _object_prefix.

jiechenbo commented on ext/storage_sources/gcp_store/gcp_connection.cpp:
```diff
@@ -30,6 +30,8 @@
 
 #include <fstream>
 
+namespace gcs = ::google::cloud::storage;
```
let's put using namespace gsc = ::google::cloud::storage here, so we can remove the "using namespace" below.

sonar3ddy replied:
addressed here: 0adccbf3c97d00eb573380c9a26490d45ef417dc

jiechenbo commented on ext/storage_sources/gcp_store/gcp_connection.cpp:
```diff
@@ -53,6 +55,15 @@ gcp_connection::put_object(const std::string &object_key, const std::string &fil
+        client.DeleteObject(_bucket_name, object_key);
+
+    if (!status.ok()) {
+        std::cout << status.code();
```
I would pipe it to std::err and place a std::endl to denote place a new line too.

sonar3ddy replied:
addressed here: 0adccbf3c97d00eb573380c9a26490d45ef417dc

jiechenbo commented on ext/storage_sources/gcp_store/gcp_connection.cpp:
```diff
@@ -53,6 +55,14 @@ gcp_connection::put_object(const std::string &object_key, const std::string &fil
 gcp_connection::delete_object(const std::string &object_key) const
 {
+    using namespace gcs;
+    gcs::Client client = _gcp_client;
```
I know we have compiling errors for this if we remove the assignment. Let's investigate why we can't directly use _gcp_client.

sonar3ddy replied:
fixed

jiechenbo commented on ext/storage_sources/gcp_store/gcp_connection.cpp:
```diff
@@ -53,6 +55,14 @@ gcp_connection::put_object(const std::string &object_key, const std::string &fil
 {
+    using namespace gcs;
+    gcs::Client client = _gcp_client;
+    google::cloud::Status status = client.DeleteObject(_bucket_name, object_key);
```
we can use auto here.

sonar3ddy replied:
addressed here: 0adccbf3c97d00eb573380c9a26490d45ef417dc

jiechenbo commented on ext/storage_sources/gcp_store/test/test_gcp_connection.cpp:
```diff
@@ -31,4 +31,13 @@
 #include "gcp_connection.h"
 
-TEST_CASE("Testing class gcpConnection", "gcp-connection") {}
+TEST_CASE("Testing class gcpConnection", "gcp-connection")
```
Let's remove this test case since putObject is not implemented yet.

sonar3ddy replied:
fixed

jiechenbo commented on ext/storage_sources/gcp_store/gcp_connection.cpp:
```diff
@@ -37,37 +37,42 @@ gcp_connection::gcp_connection(const std::string &bucket_name)
 // Builds a list of object names from the bucket.
 int
-gcp_connection::list_objects(std::vector<std::string> &objects) const
+gcp_connection::list_objects(std::vector<std::string> &objects)
```
let's place back the const for functions other than delete object

sonar3ddy replied:
fixed

jiechenbo reviewed (approved):
LGTM! Please add a comment describing changes done in the PR before getting andrew to review!

jiechenbo commented:
Was there a reason why we changed the object_key to object_name?

sonar3ddy commented:
> Was there a reason why we changed the object_key to object_name?

I changed it to match gcp function argument names. I have changed it back to object_key for consistency.

ajmorton reviewed (approved):
LGTM 👍 
We'll need to replace the `std::cerr` calls once WT-10368 is delivered but I'm happy to deal with that in WT-10368.

ajmorton commented:
@sonar3ddy just to be confirm are we pushing all testing to a different ticket? I assume we are based on [the conversation in the WT-10348 PR](https://github.com/wiredtiger/wiredtiger/pull/8717#discussion_r1073030220)

sonar3ddy commented:
> @sonar3ddy just to be confirm are we pushing unit testing to a different ticket? I assume we are based on [the conversation in the WT-10348 PR](https://github.com/wiredtiger/wiredtiger/pull/8717#discussion_r1073030220)

Yes, unit testing will be in a different ticket.

jiechenbo reviewed (approved):
LGTM! Please fix the title and ticket name to delete_object() instead

ajmorton reviewed (approved):
LGTM 🎉

sonar3ddy commented:
evergreen merge
